Add no_std + alloc feature configuration

lopopolo opened the pull request:
This commit enables `bstr` to build and test without a dependency on
`std`. This change was mostly munging `use` statements to prefer `core`
and `alloc` and changing around `cfg`s.

This change also enables `bstr` to successfully test in `no_std` and
`alloc` configurations. The following succeed where they did not even
compile before:

```
cargo test --no-default-features --lib
cargo test --no-default-features --lib --features alloc
```

Fixes #79.

lopopolo commented on src/lib.rs:
```diff
@@ -366,17 +366,20 @@ to write correct code for Unix, at the cost of getting a corner case wrong on
 */
 
-#![cfg_attr(not(feature = "std"), no_std)]
+#![no_std]
```
This change requires all items in the prelude to be imported even when linking `std`. It made fixing all of the import errors much easier. I also saw a recommendation in an issue on the Rust issue tracker that preferred this syntax. I think I've seen it used in `wasm-bindgen`.

BurntSushi replied:
Hmmm, this is not how I've done it in the past. Could you say more about what becomes easier with this?

One thing that this makes kind of a pain is that (AIUI) we can't benefit from std's prelude in tests if we do this, right? That's kind of a bummer and I'd rather have the prelude available in tests.

lopopolo replied:
I found the docs that refer to this approach:

https://github.com/rust-lang/cargo/blob/58a961314437258065e23cb6316dfc121d96fb71/src/doc/src/reference/features-examples.md#no_std-support

thomcc replied:
> One thing that this makes kind of a pain is that (AIUI) we can't benefit from std's prelude in tests if we do this, right? That's kind of a bummer and I'd rather have the prelude available in tests.

You can do this too:
```
#![cfg_attr(not(test), no_std)]
#[cfg(feature = "std")]
extern crate std;
```

(or `#[cfg(all(not(test), feature = "std"))]` on the `extern crate std` — I don't think there's a difference)

But given that bstr already has most of the tests in `mod tests` you could just put `use std::prelude::v1::*` at the top of it and get most of the benefit.

It's kind of annoying that there's not a better way to do this — the way the prelude works for no_std is janky, honestly.

lopopolo commented:
I wanted to note that none of the examples compile without default features since they depend on the `std` feature. I think the way to fix this is add `[[example]]` stanzas to `Cargo.toml` with `required-features = "std"`, but I considered this too invasive for this PR so decided against doing it myself.

lopopolo commented on src/lib.rs:
```diff
@@ -366,17 +366,20 @@ to write correct code for Unix, at the cost of getting a corner case wrong on
+extern crate alloc;
+
 #[cfg(feature = "std")]
-extern crate core;
+extern crate std;
```
If you'd prefer, I can make this `#[cfg(any(test, feature = "std"))]` and remove the `alloc` gates from tests.

BurntSushi replied:
Yes, I would like that. Otherwise, I think writing tests becomes too annoying. With that said, I do think that things like `cargo t --no-default-features` and `cargo t --no-default-features --features alloc` should at least compile fine (and run any relevant tests, if any, probably none). So I think most tests do need to be gated on `#[cfg(feature = "std")]`.

programmerjake commented:
ping -- I'd like to use bstr with alloc but without std -- I'm trying to use BString kinda like OsString as the backing type for a POSIX Path type for a virtualized filesystem I'm writing.

lopopolo commented:
@BurntSushi Friendly poke. Are you interested in this patch?

BurntSushi reviewed (changes requested):
Thanks! I left a couple of minor comments.

I'm of course interested in this PR, but me taking months to get bigger PRs merged is unfortunately just reality. I maintain a lot of projects and it is very costly for me to be constantly context switching.

One thing I noticed in this diff is that `use` statements have gotten a bit messier. Ideally, we'd switch over to nested `use` statements, but I think this diff is big enough already. With that said, one improvement would be to consistently keep `use` statements in at least three distinct blocks: `core/alloc/std`, then third-party crate imports and then finally crate-internal imports. In some parts of this PR, I see the first two categories lumped together.

I think some kind of docs are also needed here. Probably at least in the README where the crate features are documented.

lopopolo commented:
hey @BurntSushi, I know you have a lot of surface area on GitHub. This is a friendly poke that I've addressed all of your feedback on this PR and rebased it to the latest master.
